### PR TITLE
[Jormungandr] Feed attribute occupancy in stop_schedules.date_time from /vehicle_occupancies 

### PR DESCRIPTION
**Pull request by kadhikari**
concerned ticket: https://jira.kisio.org/browse/NAVP-1749

### CODE REVIEW

File: source/jormungandr/jormungandr/external_services/vehicle_occupancy.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        self.service_url = service_url
+        self.timeout = timeout
+        self.breaker = pybreaker.CircuitBreaker(
+            fail_max=kwargs.get('circuit_breaker_max_fail', app.config['CIRCUIT_BREAKER_MAX_FORSETI_FAIL']),
+            reset_timeout=kwargs.get(
+                'circuit_breaker_reset_timeout', app.config['CIRCUIT_BREAKER_FORSETI_TIMEOUT_S']
+            ),
+        )
+
+    @cache.memoize(app.config.get(str('default_cache'), {}).get(str('TIMEOUT_FORSETI'), 60 * 60))
```
Comment (xlqian) [inline source/jormungandr/jormungandr/external_services/vehicle_occupancy.py]: please use `CACHE_CONFIGURATION`, `default_cache` is just an example of cache. `CACHE_CONFIGURATION` is used to deploy.
IMO, a better way to get the default_cache is:
```suggestion
    @cache.memoize(app.config.get(str('CACHE_CONFIGURATION'), app.config.get(str('default_cache')).get(str('TIMEOUT_FORSETI'), 60 * 60))
```
But it's too verbose...

File: source/jormungandr/jormungandr/schedule.py [hunk 2]
```diff
@@ -367,6 +367,16 @@ def worker(rt_proxy, route_point, request, stop_schedule):
             rt_proxy, schedule, next_rt_passages = future.get()
             rt_proxy._update_stop_schedule(schedule, next_rt_passages)
 
+    def _manage_occupancies(self, schedules):
+        vo_service = self.instance.external_service_provider_manager.get_vehicle_occupancy_service()
+        if vo_service:
```
Comment (xlqian) [inline source/jormungandr/jormungandr/schedule.py]: plz reverse the `if` condition with `if not vo_service`

File: source/jormungandr/jormungandr/schedule.py [hunk 3]
```diff
@@ -367,6 +367,16 @@ def worker(rt_proxy, route_point, request, stop_schedule):
...
+            for schedule in schedules:
+                sp_id = schedule.stop_point.uri
+                for date_time in schedule.date_times:
+                    vj_id = date_time.properties.vehicle_journey_id
+                    args = {"stop_id": sp_id, "vehiclejourney_id": vj_id}
+                    date_time.occupancy = vo_service.get_response(args)
```
Comment (xlqian) [inline source/jormungandr/jormungandr/schedule.py]: TODO: this call may be parallelized with gevent

File: source/jormungandr/jormungandr/external_services/external_service.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        if response.status_code == 503:
+            raise FreeFloatingUnavailable('forseti responded with 503')
+        if response.status_code != 200:
+            error_msg = 'free-floating request failed with HTTP code {}'.format(response.status_code)
+            if response.text:
+                error_msg += ' ({})'.format(response.text)
+            raise FreeFloatingError(error_msg)
+
+
+class FreeFloatingError(RuntimeError):
```
Comment (xlqian) [inline source/jormungandr/jormungandr/external_services/external_service.py]: minor, can you move these errors to the very beginning of this file so errors can be grouped ?

File: source/jormungandr/jormungandr/external_services/external_service.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    def get_response(self, arguments):
+        """
+        Get free-floating information from Forseti webservice
+        """
+        pass
+
+    @classmethod
+    def _check_response(cls, response):
+        if response is None:
+            raise FreeFloatingError('impossible to access free-floating service')
```
Comment (benoit-bst) [inline source/jormungandr/jormungandr/external_services/external_service.py]: Why in external_service file you handle free_floating error ?
External service is not an upper class without specification ?

File: source/jormungandr/jormungandr/external_services/free_floating.py [hunk 6]
```diff
@@ -52,59 +53,14 @@ def __init__(self, service_url, timeout=2, **kwargs):
...
-        if response.status_code == 503:
-            raise FreeFloatingUnavailable('forseti responded with 503')
-        if response.status_code != 200:
-            error_msg = 'free-floating request failed with HTTP code {}'.format(response.status_code)
-            if response.text:
-                error_msg += ' ({})'.format(response.text)
-            raise FreeFloatingError(error_msg)
-
     @classmethod
     def response_marshaler(cls, response):
```
Comment (benoit-bst) [inline source/jormungandr/jormungandr/external_services/free_floating.py:109]: Marshaller ?

**Review by xlqian — APPROVED** [review]

File: source/jormungandr/jormungandr/external_services/vehicle_occupancy.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        """
+        Get vehicle_occupancy information from Forseti webservice
+        """
+        raw_response = self._call_webservice(arguments)
+        resp = self.response_marshaller(raw_response)
+        vehicle_occupancies = resp.get('vehicle_occupancies', [])
+        if len(vehicle_occupancies) > 0:
+            vo = vehicle_occupancies[0]['occupancy']
+            return vo
+        return 0
```
Comment (azime) [inline source/jormungandr/jormungandr/external_services/vehicle_occupancy.py]: Minor:
```suggestion
        return vehicle_occupancies[0].get('occupancy', 0) if vehicle_occupancies else 0
```

File: source/jormungandr/jormungandr/external_services/external_service.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    @abc.abstractmethod
+    def get_response(self, arguments):
+        """
+        Get external service information from Forseti webservice
+        """
+        pass
+
+    @classmethod
+    def _check_response(cls, response):
+        if response is None:
```
Comment (azime) [inline source/jormungandr/jormungandr/external_services/external_service.py]: Minor:
```suggestion
        if not response:
```

**Comment by azime** [comment]
OK for me :+1:

**Review by azime — APPROVED** [review]

**Pull request merged**